node/core/prospective-parachains/src/lib.rs

```diff
@@ -0,0 +1,500 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Implementation of the Prospective Parachains subsystem - this tracks and handles
+//! prospective parachain fragments and informs other backing-stage subsystems
+//! of work to be done.
+//!
+//! This is the main coordinator of work within the node for the collation and
+//! backing phases of parachain consensus.
+//!
+//! This is primarily an implementation of "Fragment Trees", as described in
+//! [`polkadot_node_subsystem_util::inclusion_emulator::staging`].
+//!
+//! This also handles concerns such as the relay-chain being forkful,
+//! session changes, predicting validator group assignments.
+
+use std::collections::{HashMap, HashSet};
+
+use futures::{channel::oneshot, prelude::*};
+
+use polkadot_node_subsystem::{
+	messages::ChainApiMessage, overseer, ActiveLeavesUpdate, FromOverseer, OverseerSignal,
+	SubsystemContext,
+};
+use polkadot_node_subsystem_util::inclusion_emulator::staging::{Constraints, RelayChainBlockInfo};
+use polkadot_primitives::vstaging::{
+	CandidateHash, CommittedCandidateReceipt, Hash, Id as ParaId, PersistedValidationData,
+};
+
+use crate::{
+	error::{FatalError, FatalResult, JfyiError, JfyiErrorResult, Result},
+	fragment_tree::{CandidateStorage, FragmentTree, Scope as TreeScope},
+};
+
+mod error;
+mod fragment_tree;
+
+const LOG_TARGET: &str = "parachain::prospective-parachains";
+
+// The maximum depth the subsystem will allow. 'depth' is defined as the
+// amount of blocks between the para head in a relay-chain block's state
+// and a candidate with a particular relay-parent.
+//
+// This value is chosen mostly for reasons of resource-limitation.
+// Without it, a malicious validator group could create arbitrarily long,
+// useless prospective parachains and DoS honest nodes.
+const MAX_DEPTH: usize = 4;
+
+// The maximum ancestry we support.
+const MAX_ANCESTRY: usize = 5;
+
+// TODO [now]: add this enum to the broader subsystem types.
+/// Messages sent to the Prospective Parachains subsystem.
+pub enum ProspectiveParachainsMessage {
+	/// Inform the Prospective Parachains Subsystem of a new candidate.
+	CandidateSeconded(ParaId, CommittedCandidateReceipt, PersistedValidationData),
+	/// Inform the Prospective Parachains Subsystem that a previously seconded candidate
+	/// has been backed. This requires that `CandidateSeconded` was sent for the candidate
+	/// some time in the past.
```

Comment:
> This requires that `CandidateSeconded` was sent for the candidate some time in the past

This is a precondition to be met by the caller, but it seems reasonable in practice unless we started persisting statements to the DB.

Reply:
There's not much point in re-populating the backing subsystem from disk when we restart because the relay-chain has moved on already. And if it hasn't, our peers in statement distribution will catch us up to all the candidates, and we'll follow the preconditions of `CandidateSeconded` having been sent first. If we did re-populate, we'd have to take steps to ensure these preconditions were met anyway. But why would we?